### PR TITLE
Optimize FileWriteOutBytes to avoid high system cpu usage

### LINKED ISSUE #9721: Frequent calls to FileWriteOutBytes.size() will result in high sys CPU
### Affected Version

0.13.0+

### Description

In order to compare the disk performance between local disk and cloud disk, we replace the ```OffHeapMemorySegmentWriteOutMediumFactory``` with ```TmpFileSegmentWriteOutMediumFactory``` when instancing **INDEX_MERGER_V9** in IndexMergeBenchmark(rename to IndexMergeWithTmpFileBenchmark). However, during benchmark running, we found that the sys cpu is too high:

image

image

With the help of flame graph, we found that every time we call the size() method will trigger flush() method which call write system call. 

To avoid calling flush method, we introduce a new variable ```writeOutBytes``` to record the number of bytes written:

```
final class FileWriteOutBytes extends WriteOutBytes
{
...
  private long writeOutBytes;
...
  FileWriteOutBytes(File file, FileChannel ch)
  {
    this.file = file;
    this.ch = ch;
    this.writeOutBytes = 0L;
  }
...  
  @Override
  public void write(int b) throws IOException
  {
    flushIfNeeded(1);
    buffer.put((byte) b);
    writeOutBytes++;
  }

  @Override
  public void writeInt(int v) throws IOException
  {
    flushIfNeeded(Integer.BYTES);
    buffer.putInt(v);
    writeOutBytes += Integer.BYTES;
  }

  @Override
  public int write(ByteBuffer src) throws IOException
  {
...
    buffer.put(src);
    writeOutBytes += len;
    return len;
  }

  @Override
  public long size() throws IOException
  {
    return writeOutBytes;
  }
```

Then we rerun the benchmark and sys cpu returned to normal. The benchmark report shows that the performance is improved by about 44%.

* Machine info: 
CPU: Intel(R) Xeon(R) Gold 5118 CPU @ 2.30GHz
Disk: HDD
Command: java -Djava.io.tmpdir=/data00/tmp_dir -jar benchmarks.jar IndexMergeWithTmpFileBenchmark

before optimization:
Benchmark                               (numSegments)  (rollup)  (rowsPerSegment)  (schema)  Mode  Cnt        Score       Error  Units
IndexMergeWithTmpFileBenchmark.mergeV9              5      true             75000     basic  avgt   25  8161891.327 ± 32636.767  us/op
IndexMergeWithTmpFileBenchmark.mergeV9              5     false             75000     basic  avgt   25  8041137.131 ± 41477.861  us/op

after optimization:
Benchmark                               (numSegments)  (rollup)  (rowsPerSegment)  (schema)  Mode  Cnt        Score       Error  Units
IndexMergeWithTmpFileBenchmark.mergeV9              5      true             75000     basic  avgt   25  4536098.486 ± 13668.764  us/op
IndexMergeWithTmpFileBenchmark.mergeV9              5     false             75000     basic  avgt   25  4321243.165 ± 30293.772  us/op

### PR DESCRIPTION
**Pull request by viongpanzi**
Fixes #9721.

### Description

see #9721

The benchmark report shows that the performance is improved by about 44% with this pr.

Machine info:
CPU: Intel(R) Xeon(R) Gold 5118 CPU @ 2.30GHz
Disk: HDD
Command: java -Djava.io.tmpdir=/data00/tmp_dir -jar benchmarks.jar IndexMergeWithTmpFileBenchmark

```
before optimization:
Benchmark (numSegments) (rollup) (rowsPerSegment) (schema) Mode Cnt Score Error Units
IndexMergeWithTmpFileBenchmark.mergeV9 5 true 75000 basic avgt 25 8161891.327 ± 32636.767 us/op
IndexMergeWithTmpFileBenchmark.mergeV9 5 false 75000 basic avgt 25 8041137.131 ± 41477.861 us/op

after optimization:
Benchmark (numSegments) (rollup) (rowsPerSegment) (schema) Mode Cnt Score Error Units
IndexMergeWithTmpFileBenchmark.mergeV9 5 true 75000 basic avgt 25 4536098.486 ± 13668.764 us/op
IndexMergeWithTmpFileBenchmark.mergeV9 5 false 75000 basic avgt 25 4321243.165 ± 30293.772 us/op
```

<hr>

This PR has:
- [x] been self-reviewed.
- [x] been tested in a test Druid cluster.

<hr>

##### Key changed/added classes in this PR
 * `FileWriteOutBytes`
 * `IndexMergeWithTmpFileBenchmark`

### CODE REVIEW

File: processing/src/main/java/org/apache/druid/segment/writeout/FileWriteOutBytes.java [hunk 1]
```diff
@@ -105,8 +110,8 @@ public void write(byte[] b, int off, int len) throws IOException
   @Override
   public long size() throws IOException
   {
-    flush();
-    return ch.size();
+    flushIfNeeded(0); // To avoid check the declared IOException never thrown error
```
Comment (suneet-s) [inline processing/src/main/java/org/apache/druid/segment/writeout/FileWriteOutBytes.java:113]: I think this is cleaner

```
  public long size()
  {
    return writeOutBytes;
  }
```

**Comment by suneet-s** [comment]
Thanks for the fix and the benchmarks! I haven't read through the benchmark changes but it makes sense why this change would have a big perf benefit.

The only thought I had here is - how do we make sure that writeOutBytes is always in sync. I couldn't find any unit tests for FileWriteOutBytes, but is it covered by some other test?

**Comment by suneet-s** [comment]
I checked out the PR locally and ran tests to see the coverage. Line coverage is quite high and it's hit by a lot of different tests.
Screen Shot 2020-04-21 at 1 34 24 PM

File: processing/src/main/java/org/apache/druid/segment/writeout/FileWriteOutBytes.java (hunk 1 above)
Comment (viongpanzi) [reply processing/src/main/java/org/apache/druid/segment/writeout/FileWriteOutBytes.java:113]: Thanks for point out that. I added this line of code on purpose, or `The declared exception <code>IOException</code> is never thrown` error will block ci. And I attempt to remove `throws IOException` from method signature, but new error occurs! Do you have any suggestions?

**Comment by viongpanzi** [comment]
> Thanks for the fix and the benchmarks! I haven't read through the benchmark changes but it makes sense why this change would have a big perf benefit.
> 
> The only thought I had here is - how do we make sure that writeOutBytes is always in sync. I couldn't find any unit tests for FileWriteOutBytes, but is it covered by some other test?

Good question！There are some unit tests in `WriteOutBytesTest.java`, so I do not add new test cases. I'll try to add more test cases.

**Comment by viongpanzi** [comment]
> I've written tests on top of this change and caught one bug - would you incorporate my tests and fix into this PR please.
> 
> https://github.com/suneet-s/druid/blob/69954723bddf223ae1eddb5387de10bcd88b8e7a/processing/src/test/java/org/apache/druid/segment/writeout/FileWriteOutBytesTest.java#L99-L121
> 
> If it makes it easier to copy the tests and bug fix, I can make a PR from that branch.
> 
> Other than that bug, the change looks good.

Nice job! Thanks @suneet-s . I think a PR from that branch is better. `FileWriteOutBytesTest.java` is added after this pr and I will update this pr today later.

**Comment by suneet-s** [comment]
> > I've written tests on top of this change and caught one bug - would you incorporate my tests and fix into this PR please.
> > https://github.com/suneet-s/druid/blob/69954723bddf223ae1eddb5387de10bcd88b8e7a/processing/src/test/java/org/apache/druid/segment/writeout/FileWriteOutBytesTest.java#L99-L121
> > If it makes it easier to copy the tests and bug fix, I can make a PR from that branch.
> > Other than that bug, the change looks good.
> 
> Nice job! Thanks @suneet-s . I think a PR from that branch is better. `FileWriteOutBytesTest.java` is added after this pr and I will update this pr today later.

@viongpanzi Here you go - #9741

You can run `git fetch upstream && git merge upstream/master` to pull in the test class FileWriteOutBytesTest

**Comment by viongpanzi** [comment]
> > > I've written tests on top of this change and caught one bug - would you incorporate my tests and fix into this PR please.
> > > https://github.com/suneet-s/druid/blob/69954723bddf223ae1eddb5387de10bcd88b8e7a/processing/src/test/java/org/apache/druid/segment/writeout/FileWriteOutBytesTest.java#L99-L121
> > > If it makes it easier to copy the tests and bug fix, I can make a PR from that branch.
> > > Other than that bug, the change looks good.
> > 
> > 
> > Nice job! Thanks @suneet-s . I think a PR from that branch is better. `FileWriteOutBytesTest.java` is added after this pr and I will update this pr today later.
> 
> @viongpanzi Here you go - #9741
> 
> You can run `git fetch upstream && git merge upstream/master` to pull in the test class FileWriteOutBytesTest

Thanks @suneet-s . PR updated!

**Comment by suneet-s** [comment]
@viongpanzi I don't see the bug fix or the tests updated in this pr.

To copy the fix and tests into this pr you can run

```
git fetch upstream pull/9741/head:pr_9741 && git merge pr_9741
```

and then push the patch back up to github

**Comment by viongpanzi** [comment]
> @viongpanzi I don't see the bug fix or the tests updated in this pr.
> 
> To copy the fix and tests into this pr you can run
> 
> ```
> git fetch upstream pull/9741/head:pr_9741 && git merge pr_9741
> ```
> 
> and then push the patch back up to github

@suneet-s Sorry! Now the tests from 9741 are merged!

**Comment by suneet-s** [comment]
@viongpanzi Looks like the intelliJ inspections failure is valid (I now understand why you left that comment in the first patch.

The correct fix here is to change the abstract method definition in WriteOutBytes to

```
public abstract long size();
```

since none of the implementations throw an IOException anymore when fetching the size.

Can you please fix this so this fix can be merged

**Comment by viongpanzi** [comment]
> @viongpanzi Looks like the intelliJ inspections failure is valid (I now understand why you left that comment in the first patch.
> 
> The correct fix here is to change the abstract method definition in WriteOutBytes to
> 
> ```
> public abstract long size();
> ```
> 
> since none of the implementations throw an IOException anymore when fetching the size.
> 
> Can you please fix this so this fix can be merged

@suneet-s Updated!

**Comment by suneet-s** [comment]
@viongpanzi - The intelliJ inspection errors still exist in this patch. You can run intelliJ inspections locally from the editor like this.
Screen Shot 2020-04-22 at 1 44 50 PM
The failing inspection was `Redundant 'throws' clause`

I've fixed this in #9748 - Please pull in that fix into this PR using

```
git fetch upstream pull/9748/head:pr_9748 && git merge pr_9748
```
and push back to github. This time travis should succeed 🤞

**Comment by viongpanzi** [comment]
> @viongpanzi - The intelliJ inspection errors still exist in this patch. You can run intelliJ inspections locally from the editor like this.
> Screen Shot 2020-04-22 at 1 44 50 PM
> The failing inspection was `Redundant 'throws' clause`
> 
> I've fixed this in #9748 - Please pull in that fix into this PR using
> 
> ```
> git fetch upstream pull/9748/head:pr_9748 && git merge pr_9748
> ```
> 
> and push back to github. This time travis should succeed 🤞

That's great! Thanks @suneet-s ! I didn't know that before. I've run through the test locally. 
The PR is updated.

**Pull request merged**